src/protocol/boolean/random_bits_generator.rs

```diff
@@ -93,110 +68,102 @@ where
     /// inner members multiple times, I/O errors while executing MPC protocols,
     /// read from an empty buffer, etc.
     pub async fn generate(&self, record_id: RecordId) -> Result<RandomBitsShare<F, S>, Error> {
-        let share = if let Some(v) = self.default_generator.next(Some(record_id)).await? {
+        let share = if let Some(v) = solved_bits(self.ctx.clone(), record_id).await? {
             v
         } else {
             loop {
-                self.abort_count.fetch_add(1, Ordering::AcqRel);
-                if let Some(v) = self.fallback_generator.next(None).await? {
+                let i = self.fallback_count.fetch_add(1, Ordering::AcqRel);
+                if let Some(v) = solved_bits(self.fallback_ctx.clone(), RecordId::from(i)).await? {
                     break v;
                 }
             }
         };
 
-        if self.total_records.last(record_id) {
-            self.fallback_generator.close().await;
+        if self.ctx.is_last_record(record_id) {
+            // TODO: close indeterminate channels
```

Comment:
I'm still not sure how to address this one.

What I was working toward, which might not work out, is to have the context carry an enum:

```rust
enum RecordCloser {
  KnownRecordCount(usize),
  IndeterminateRecordCount(Step),
}
```

The `Step` type here would be a step that you use as a key into an `Arc<Mutex<HashMap<Step, Vec<GatewaySender>>>>` (or a `DashMap` along similar lines) that is owned by the context (in the `inner` part, that is).

The idea is that when you call `set_total_records(TotalRecords::Indeterminate)` a new item is added to the store and whenever a new channel is created from a context that has this `RecordCloser::IndeterminateRecordCount(s)`, a copy of the sender is saved to the identified store in the collection.

Any code that creates an indeterminate step would be responsible for closing all the senders that are gathered into this collection when they are done, using the top-level step to identify which ones need closing.

Reply:
Note that I don't think we should over-index on trying to reduce contention on this store.  We expect low utilization.  The only thing we really need to do this for is `RandomBitsGenerator`, which has to be constructed just once by design.  Therefore, we can be fairly confident that adding the item to the context store is sufficiently rare and opening channels is also rare.  Starting with something crude should do, which might mean not using `DashMap` or similar.

Reply:
> The idea is that when you call set_total_records(TotalRecords::Indeterminate) a new item is added to the store and whenever a new channel is created from a context that has this RecordCloser::IndeterminateRecordCount(s), a copy of the sender is saved to the identified store in the collection.

I had the same thought, but it seemed to me that we are trying to optimize the rare case and being explicit here shouldn't be expensive. Sure, RBG is unique, let it explicitly close the fallback channel.

Reply:
> we seem to require it in more than one protocol and the current design requires callers to provide record_id, which means we need an instance per step?

You currently need a single RBG for all records on a step, though you can create a different RBG for a different step.

Reply:
> RBG is unique, let it explicitly close the fallback channel.

So that is hard.  If you look at RBG, it spawns multiple steps, even for fallback.  So what we need is to close all of those channels.  Hence the suggested design.  As noted, this should be rare, so we don't need lots of machinery, or very fast code, but it's a little more complex than just closing a single channel.